deepinv/physics/generator/mri.py

```diff
@@ -1,83 +1,244 @@
+        )
 
-        if len(self.img_size) == 2:
-            H, W = self.img_size
-            C = 2
-        elif len(self.img_size) == 3:
-            C, H, W = self.img_size
-        else:
-            raise ValueError("img_size must be (C, H, W) or (H, W)")
-
-        acceleration_factor = self.acceleration
-
-        if acceleration_factor == 4:
-            central_lines_percent = 0.08
-            num_lines_center = int(central_lines_percent * W)
-            side_lines_percent = 0.25 - central_lines_percent
-            num_lines_side = int(side_lines_percent * W)
-        if acceleration_factor == 8:
-            central_lines_percent = 0.04
-            num_lines_center = int(central_lines_percent * W)
-            side_lines_percent = 0.125 - central_lines_percent
-            num_lines_side = int(side_lines_percent * W)
-
-        mask = torch.zeros((batch_size, H, W), **self.factory_kwargs)
-
-        center_line_indices = torch.linspace(
-            H // 2 - num_lines_center // 2,
-            H // 2 + num_lines_center // 2 + 1,
-            steps=50,
-            dtype=torch.long,
+        if self.T == 0:
+            mask = mask[:, :, 0, :, :]
+
+        return {"mask": mask}
+
+
+class RandomMaskGenerator(BaseMaskGenerator):
+    """Generator for MRI Cartesian acceleration masks using random uniform undersampling.
+
+    Generate a mask of vertical lines for MRI acceleration with fixed sampling in low frequencies (center of k-space) and random uniform undersampling in the high frequencies.
+
+    Supports k-t sampling, where the mask is selected randomly across time.
+
+    The mask is repeated across channels and randomly varies across batch dimension.
+
+    For parameter descriptions see :class:`deepinv.physics.generator.mri.BaseMaskGenerator`
+
+    Example:
+
+    ::
+
+        generator = RandomMaskGenerator((2, 8, 64, 64), acceleration=8, center_fraction=0.04) # C, T, H, W
+        params = generator.step(batch_size=1)
+        mask = params["mask"]
+        mask.shape # 1, 2, 8, 64, 64
+
+    """
+
+    def get_pdf(self) -> torch.Tensor:
+        """Create one-dimensional uniform probability density function across columns, ignoring any fixed sampling columns.
+
+        :return torch.Tensor: unnormalised 1D vector representing pdf evaluated across mask columns.
+        """
+        return np.ones(self.W)
+
+    def sample_mask(self, mask: torch.Tensor) -> torch.Tensor:
+        pdf = self.get_pdf()
+
+        # lines are never randomly sampled from the already sampled center
+        pdf[
+            self.W // 2 - self.n_center // 2 : self.W // 2 + ceildiv(self.n_center, 2)
+        ] = 0
+
+        # normalise distribution
+        pdf /= np.sum(pdf)
+
+        # select low-frequency lines according to pdf
+        for b in range(mask.shape[0]):
+            for t in range(mask.shape[2]):
+                idx = self.rng.choice(self.W, self.n_lines, replace=False, p=pdf)
```

Comment:
I have the feeling that this is going to be a bit slow for large batch sizes

Can we use torch.randperm(self.W)[:self.n_lines] ?

See [this post](https://stackoverflow.com/questions/59461811/random-choice-with-pytorch)

Reply:
This might be tricky - you would have to use `torch.multinomial(pdf, n_lines, replacement=False)` - but this still won't help with vectorising the loop, which I think is needed as you want the acceleration rate to be same each time-step and each image in batch dimension.

Reply:
ok, we can keep it like this and maybe optimise in the future.